File: app/csharp/AccountingSystem.Application/Services/JournalEntryEventSourcingService.cs
namespace AccountingSystem.Application.Services;

using AccountingSystem.Application.Ports.In;
using AccountingSystem.Application.Ports.Out;
using AccountingSystem.Domain.Aggregates;

/// <summary>
/// 仕訳イベントソーシングサービス
/// </summary>
public class JournalEntryEventSourcingService : IJournalEntryEventSourcingService
{
    private readonly IEventStoreRepository _eventStoreRepository;

    public JournalEntryEventSourcingService(IEventStoreRepository eventStoreRepository)
    {
        _eventStoreRepository = eventStoreRepository;
    }

    /// <summary>
    /// 仕訳を作成
    /// </summary>
    public async Task<string> CreateJournalEntryAsync(
        DateOnly entryDate,
        string description,
        List<LineItemDto> lineItems,
        string userId)
    {
        var id = Guid.NewGuid().ToString();

        // Aggregate 作成（コマンド実行）
        var aggregate = JournalEntryAggregate.Create(
            id,
            entryDate,
            description,
            lineItems.Select(dto => new JournalEntryAggregate.LineItem(
                dto.AccountCode,
                Enum.Parse<JournalEntryAggregate.DebitCredit>(dto.DebitCredit),
                dto.Amount
            )).ToList(),
            userId
        );

        // イベントをイベントストアに保存
        await _eventStoreRepository.SaveAsync(
            id,
            aggregate.UncommittedEvents,
            0  // 新規作成なので expectedVersion = 0
        );

        aggregate.MarkEventsAsCommitted();

        return id;
    }

    /// <summary>
    /// 仕訳を承認
    /// </summary>
    public async Task ApproveJournalEntryAsync(string journalEntryId, string approvedBy, string comment)
    {
        // イベント再生で現在の状態を復元
        var events = await _eventStoreRepository.GetEventsAsync(journalEntryId);
        if (events.Count == 0)
        {
            throw new ArgumentException($"仕訳が見つかりません: {journalEntryId}");
        }

        var aggregate = JournalEntryAggregate.Replay(events);
        var currentVersion = aggregate.Version;

        // コマンド実行
        aggregate.Approve(approvedBy, comment);

        // 新しいイベントを保存
        await _eventStoreRepository.SaveAsync(
            journalEntryId,
            aggregate.UncommittedEvents,
            currentVersion
        );

        aggregate.MarkEventsAsCommitted();
    }

    /// <summary>
    /// 仕訳を削除
    /// </summary>
    public async Task DeleteJournalEntryAsync(string journalEntryId, string reason, string userId)
    {
        var events = await _eventStoreRepository.GetEventsAsync(journalEntryId);
        if (events.Count == 0)
        {
            throw new ArgumentException($"仕訳が見つかりません: {journalEntryId}");
        }

        var aggregate = JournalEntryAggregate.Replay(events);
        var currentVersion = aggregate.Version;

        aggregate.Delete(reason, userId);

        await _eventStoreRepository.SaveAsync(
            journalEntryId,
            aggregate.UncommittedEvents,
            currentVersion
        );

        aggregate.MarkEventsAsCommitted();
    }

    /// <summary>
    /// 仕訳を取得（イベント再生）
    /// </summary>
    public async Task<JournalEntryAggregate> GetJournalEntryAsync(string journalEntryId)
    {
        var events = await _eventStoreRepository.GetEventsAsync(journalEntryId);
        if (events.Count == 0)
        {
            throw new ArgumentException($"仕訳が見つかりません: {journalEntryId}");
        }
        return JournalEntryAggregate.Replay(events);
    }

    /// <summary>
    /// 特定時点の仕訳状態を取得（タイムトラベル）
    /// </summary>
    public async Task<JournalEntryAggregate> GetJournalEntryAtAsync(string journalEntryId, DateTime pointInTime)
    {
        var events = await _eventStoreRepository.GetEventsUntilAsync(journalEntryId, pointInTime);
        if (events.Count == 0)
        {
            throw new ArgumentException($"指定時点の仕訳が見つかりません: {journalEntryId}");
        }
        return JournalEntryAggregate.Replay(events);
    }
}

File: app/fsharp/AccountingSystem.Application/Services/JournalEntryEventSourcingService.fs
namespace AccountingSystem.Application.Services

open System
open System.Threading.Tasks
open AccountingSystem.Domain.Events
open AccountingSystem.Domain.Aggregates
open AccountingSystem.Application.Port.In
open AccountingSystem.Application.Port.Out

/// <summary>
/// 仕訳イベントソーシングサービス
/// </summary>
type JournalEntryEventSourcingService(eventStoreRepository: IEventStoreRepository) =

    let aggregateType = "JournalEntry"

    interface IJournalEntryEventSourcingUseCase with

        /// <summary>
        /// 仕訳を作成
        /// </summary>
        member _.CreateJournalEntryAsync
            (id: string)
            (entryDate: DateTime)
            (description: string)
            (lineItems: JournalEntryLineItem list)
            (userId: string)
            : Task<Result<JournalEntryAggregate, string>> =
            task {
                // Aggregate コマンドを実行
                match JournalEntryAggregate.create id entryDate description lineItems userId with
                | Error msg -> return Error msg
                | Ok aggregate ->
                    // イベントを保存
                    let! saveResult =
                        eventStoreRepository.SaveAsync
                            id
                            aggregateType
                            aggregate.UncommittedEvents
                            0
                            userId
                            None

                    match saveResult with
                    | Error msg -> return Error msg
                    | Ok () ->
                        return Ok (JournalEntryAggregate.markEventsAsCommitted aggregate)
            }

        /// <summary>
        /// 仕訳を承認
        /// </summary>
        member _.ApproveJournalEntryAsync
            (id: string)
            (approvedBy: string)
            (approvalComment: string)
            : Task<Result<JournalEntryAggregate, string>> =
            task {
                // 現在の状態を復元
                let! events = eventStoreRepository.GetEventsAsync id

                if List.isEmpty events then
                    return Error "仕訳が見つかりません"
                else
                    let aggregate = JournalEntryAggregate.replay events

                    // Aggregate コマンドを実行
                    match JournalEntryAggregate.approve approvedBy approvalComment aggregate with
                    | Error msg -> return Error msg
                    | Ok updatedAggregate ->
                        // イベントを保存
                        let! saveResult =
                            eventStoreRepository.SaveAsync
                                id
                                aggregateType
                                updatedAggregate.UncommittedEvents
                                aggregate.Version
                                approvedBy
                                None

                        match saveResult with
                        | Error msg -> return Error msg
                        | Ok () ->
                            return Ok (JournalEntryAggregate.markEventsAsCommitted updatedAggregate)
            }

        /// <summary>
        /// 仕訳を削除
        /// </summary>
        member _.DeleteJournalEntryAsync
            (id: string)
            (reason: string)
            (userId: string)
            : Task<Result<JournalEntryAggregate, string>> =
            task {
                // 現在の状態を復元
                let! events = eventStoreRepository.GetEventsAsync id

                if List.isEmpty events then
                    return Error "仕訳が見つかりません"
                else
                    let aggregate = JournalEntryAggregate.replay events

                    // Aggregate コマンドを実行
                    match JournalEntryAggregate.delete reason userId aggregate with
                    | Error msg -> return Error msg
                    | Ok updatedAggregate ->
                        // イベントを保存
                        let! saveResult =
                            eventStoreRepository.SaveAsync
                                id
                                aggregateType
                                updatedAggregate.UncommittedEvents
                                aggregate.Version
                                userId
                                None

                        match saveResult with
                        | Error msg -> return Error msg
                        | Ok () ->
                            return Ok (JournalEntryAggregate.markEventsAsCommitted updatedAggregate)
            }

        /// <summary>
        /// 仕訳を取得（イベント再生）
        /// </summary>
        member _.GetJournalEntryAsync (id: string) : Task<JournalEntryAggregate option> =
            task {
                let! events = eventStoreRepository.GetEventsAsync id

                if List.isEmpty events then
                    return None
                else
                    let aggregate = JournalEntryAggregate.replay events
                    return Some aggregate
            }

        /// <summary>
        /// 特定時点の仕訳を取得（タイムトラベル）
        /// </summary>
        member _.GetJournalEntryAtAsync
            (id: string)
            (pointInTime: DateTime)
            : Task<JournalEntryAggregate option> =
            task {
                let! events = eventStoreRepository.GetEventsUntilAsync id pointInTime

                if List.isEmpty events then
                    return None
                else
                    let aggregate = JournalEntryAggregate.replay events
                    return Some aggregate
            }

        /// <summary>
        /// すべての仕訳を取得
        /// </summary>
        member this.GetAllJournalEntriesAsync () : Task<JournalEntryAggregate list> =
            task {
                let! aggregateIds = eventStoreRepository.GetAggregateIdsAsync aggregateType

                let! aggregates =
                    aggregateIds
                    |> List.map (fun id ->
                        task {
                            let! events = eventStoreRepository.GetEventsAsync id
                            if List.isEmpty events then
                                return None
                            else
                                let aggregate = JournalEntryAggregate.replay events
                                // 削除済みは除外
                                if aggregate.Deleted then
                                    return None
                                else
                                    return Some aggregate
                        })
                    |> Task.WhenAll

                return aggregates |> Array.choose id |> Array.toList
            }
